### ISSUE TITLE
Feature Request: Pass Response to `Retry.is_retry()`

### ISSUE THREAD

**Issue by Mr0grog** [urllib3/urllib3#1445]
I’m working on some tools for using the [Internet Archive’s Wayback](https://web.archive.org) Machine APIs and it’d be really helpful if I could write a custom `Retry` class that can look at the the response headers to determine whether a retry is warranted. In particular, when requesting the snapshot of a URL, it’s possible to get an error code that might be coming from the wayback machine or might be coming from the snapshot itself (i.e. the original URL returned the error code when it was scraped). You can determine which of these situations you’re dealing with by looking for the `Memento-Datetime` header, but [`Retry.is_retry()`](https://github.com/urllib3/urllib3/blob/f4efcca559cc753f878e0a73e28dd6e795d0d052/src/urllib3/util/retry.py#L304-L318), which seems like the right place to do this, doesn’t have access to that info.

It feels like there ought to be plenty of other odd cases where more info about the response would be helpful, so I’m wondering if passing it directly to `Retry.is_retry()` would be a good idea. Since this gets called before `Retry.increment()` adds request info to the history, passing the URL might be nice, too.

I’d love to change this signature:

```py
def is_retry(self, method, status_code, has_retry_after=False):
```

to:

```py
def is_retry(self, method, url, response):
```

(`status_code` and `has_retry_after` can both be obtained from the `response` object, and can even be handled slightly more efficiently by not asking for them before they’re needed, since `is_retry()` returns early in some cases.)

**Comment by SethMichaelLarson** [comment]
Unfortunately I'm doubtful we'll be able to change that method to remove the `status_code` in the signature. Adding `response` is more likely but it would break users that subclass the `Retry` object, I'm unsure of how often that occurs.

An easier way to accomplish this for you may be to use the Requests library along with some request-toolbelt code to handle a custom retry/redirect?

**Comment by Mr0grog** [comment]
Hmmm, it doesn’t seem like there’s anything in toolbelt for retrying or that would cover a case like this. It also feels a bit hacky to wrap my retry/backoff logic *around* `urlopen()` (or `requests.Session.send()`) when it feels like this *ought* to fit neatly in the scope of the `Retry` class. I do understand the need not to make a breaking change to public API, though.

Would it be reasonable to add `url` and `response` as optional kwargs to the end? I suppose it would be kind of ugly and involve including data that the default implementation never needs, but it would make this use case possible.

---

As a side note, wrapping retry logic around the Requests library is kind of painful because the error types it exposes are much broader, leading to not-so-good string checking like this:

```py
try:
    response = super().send(*args, **kwargs)
    if self.should_retry(response):
        # retry it
except (ProxyError, RetryError, Timeout):
    # retry it
except ConnectionError as error:
    if 'NewConnectionError' in str(error):
        # retry it
```

Not your problem here, I know. But just to point out that getting the retry logic right if wrapping Requests is not totally straightforward. 🙁

**Comment by sigmavirus24** [comment]
Yeah, there’s nothing in the tool belt for retry logic. We could add something but I have no clue what that would look like

**Comment by SethMichaelLarson** [comment]
At a minimum I'll mark this as something for 2.0.0 because I agree that we should be passing the whole response into the `Retry` object to allow for better subclassing potential.

**Comment by Mr0grog** [comment]
@SethMichaelLarson Great, thanks! I’ve put a little note in my source about 2.0 as a reminder :)

@sigmavirus24 I think if I were trying to make a nice, abstract version of this, my first thought would be a `requests.adapters.HTTPAdapter` subclass that someone could use or subclass; then they can mount that in a session as needed. It would wrap the `send()` method with retry logic.

**Comment by EnricoMi** [comment]
The following approach allows to add this feature without breaking existing `Retry` subclasses:
- Add `is_retry_response(self, method, response)` to `Retry`
- Implement it as `return self.is_retry(method, response.status, bool(response.getheader("Retry-After"))`
- Have urllib3 call into `is_retry_response` instead of `is_retry`

Users can subclass `is_retry` or `is_retry_response`.

This potentially allows to add this feature to 1.x. You can rework that in 2.x anyway.

Given this request has been opened three years ago and the proposed solutions seem to require some significant rework, an intermediate solution would be highly appreciated.

**Comment by EnricoMi** [comment]
A possible workaround **without changing urllib3** is to implement your own `Retry`, add the error code to `status_forcelist` and handle the response in `increment` instead of `is_retry`:

Here is a sketch, not actually tested code:
```
class RetryOnResponse(Retry):
    def __init__(self, **kwargs):
        # add your inspected status code here, for me it is 403
        kwargs['status_forcelist'] = kwargs.get('status_forcelist', []) + [403]
        super().__init__(**kwargs)

    # increment is being called for your status code, providing you with the response
    def increment(self,
                  method=None,
                  url=None,
                  response=None,
                  error=None,
                  _pool=None,
                  _stacktrace=None):
        if response and response.status == 403:
          # inspect response
          if not retryable:
            raise MaxRetryError(_pool, url, response.reason)

        return super().increment(method, url, response, error, _pool, _stacktrace)
```

Not pretty but works.

**Comment by EnricoMi** [comment]
@sethmlarson is it worth providing a PR for the non-breaking solution proposed in https://github.com/urllib3/urllib3/issues/1445#issuecomment-890794221?

**Comment by sethmlarson** [comment]
@EnricoMi I think so! We can discuss the intricacies with code in front of us :) Thanks!

**Comment by EnricoMi** [comment]
@sethmlarson ok, let's talk code: #2500. @Mr0grog do you think this would work for you?

**Comment by Mr0grog** [comment]
Yep, the approach in #2500 would certainly work for my case.

**Comment by Mr0grog** [comment]
Given that v2.0 shipped without this, should it be closed as not going to be fixed? (see https://github.com/urllib3/urllib3/issues/1445#issuecomment-423177707)

Or is the plan to make major version updates with possible breaking changes more often here?

**Comment by Mr0grog** [comment]
That is to say, #2500 would still be great, but it seems like the ship has sailed on the more ideal fix of changing the signature for `is_retry`.
